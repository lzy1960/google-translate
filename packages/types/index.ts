export type Options = {
  from: string
  to: string
  tld: string
}
