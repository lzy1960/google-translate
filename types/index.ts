import { LanguageKey } from './language'

export type Options = {
  from: LanguageKey
  to: LanguageKey
  tld?: string
  type?: Lowercase<RcpIdsKeys>
  isMobile?: boolean
  proxy?: string
}

export type Result = DefaultResult

export type DefaultResult = {
  from: LanguageKey
  pronunciation: string | null
  text: string
}

export enum BatchExecute {
  MOBILE = '[";MS-4L2LQAAZVbL_HdWxfNm3AgZkJywwmACkAIwj8RigqBncmLg185qOW1jp2bwl8aCZjVHIroRlAyeikYihCFiwvKh8AAAC6TwAAAAR1AQcXADnjTVI6_b56aIt6jome8mrFoy6kuqpaad5hJdcjF5vRxQwNcY47w8b8f7xehcm0VB8J9aoSoVViBHCEApY3faUPeAVmgcMzX_n9CbA3H-RAxlultCzUrkbpy69q0G_yrRjP40cmf9FKM0aYwLFYgjli4ClOZ0Cy_DZSv76JPzz-2I1TmbJe43-zBDKI9OCXTbE0Yg35VJw7cbpgf_Ls3ULUMy4_TRhAQ5N3e_cUUL6ZCht9rQqMMenCtHMzJS1EeWiKo_-Bh4-EkFJxoIOP7dCZMQfqz7hrJPP8ofKI364vYdmrW2ONFq8cz0BWbqvpRp7Pc9_npcen3s4SfDsFAVYtAqufS7lXaOnefi4pjGP1I9u6V-h4LuLDF4ZRHkMUJJkm4yU3whEz8VxRxK7jhZpHsg3MyvZImh7s-erqo3edS0Q_811LS40_1t0QfP0d91yLK9QVHy8FoMdO3rcoBZUPuu-DLBehJbIA6K1NMNZx0bEaTmbXYf_qWJ0UmyyoPL62HuLx8sSjxaVq6aZS_tCnoLrW1Vp7WV2VLS3hYm9MbPZcilzQlDx0qQZk2BVZ5HLrdn7IQ-dWUa82tyrMuriz9YgBTOh7AxTGCEgpCTG7CT1PsjbuFh0mmSCdJHqsYb0bAAurCL2j9wMIvEKmsxGlwZfwDHkY9WTLSO8VgHYFn1opCEN_9LGf0m0E83asgA8JvXh5n-vSAm2EQm7ine9C4I_czi81pAXciz3EhzcXdHPV2wLsCgSVudbetISbkqYxto_2nompwq8LGv4V5dgplac8CmPXrtkDG-svKMGlGVVd2MJmvdZ0X1L_Hz6_aoEiUY-Nqmha0P631urEH8UujRFV5Nz6lOS799rtdRU5AKappXTDckHXCrhsFkx6g2r3l5UE9pdvEvqAAQgrTeS4T6_cJ84gJR3UiSzqEl8uSHWqau87-6RHE4gPaQuoiQOiZg",null,null,426,29,null,null,0,"2"]',
  PC = '[";xdu425bQAAZVbL_HdWxfJGZfeXgQdSAmACkAIwj8Rol9p0xXYA6iT5wRunruTrduUuXi08CKpuhNTpZWMRkuxKuMAB8AAADyTwAAAAN1AQcXALNdaTASC9AXhnsFEhyF7MEg2CJBVW1ZWWYxNtZ_n9ec1DoqNpvOVOB1cxWbzu4LLnBDNPxNx9Lf5qCyh99AKzyky3ZxgiCLAha4DuBN0Ekd24bHo2IwkTCTj9ZXMFQAdOl7pzzsCM2-1MNr6U2VvRTirE7VrWeJyXypivCPwhny2QYgOTcG0FPkgz3d5xCipsXPXCfdnJ2U6hdlt39cI-XJUc0OdLNXoqw7d5elL-2F3LEWkoQCJDvwLbkHXMoWaj5PwjHr7561oC9XqMncHk5w-qHnlppii7BDTKGh5HAWfcQlqTNK4yu9y3W_uPltoFIsxx0CTW1Rkz3N11qd_oM6Sp0IvWw1vIKj3IEvRkUoCT4Hql5KRzV1FE8wDFZHtr9ghXGKH1vXkxeuI5fiGUikdcXluZnHnjI9KqcbdBwM2cG9zZmr-3fgeEnoRJkrNM8oEmurIwEhurX-mc-pqfnO6mW5xwio6_R-leTf71jDo7PY-OVwdqDm0l6bELBGXbxKylGMHdbJ1qymJcVpfC0o6U9Ru5OFtzTYaPgD3ThA6pleS2PPuD8Lrbo9oCz0Mk2KIhGIkglrub8OtzbFejo-OoLrSqXrLoRF-v7tPMpJr1Xf-tqsCT6TaKIjMF_g90jVdApbkxfeRoZ6Y5aXxf01LzAqcvDKuW_5592p9fCXeDoql2W5eBPsbzJJrgxC2cXPP2ia0dqvINlZdu2zfLIGdsZ0sSmK0aEXSomd_zs5iGq8vh4j0-QM36Fw97_sft7V5vRwQUpqBNXoeGgciZlmuPGrcNI2CB8mKXzqQOuAZmseK5KG9IipWTpo5eScBs4X5Vvr7MV_JNRm9Fw1q1j1-l2fkl51oNbs95ks_F_RNPPi4xW0PiZy9WxjP2Fvi6XP3Y04Mq8WL7sNHfJU18EFtlOgGWXd514I1wbcw02yRq8oBW15rX6qO9-vZ9PXggkBVLvDHPCtmt6a",null,null,492,26,null,null,0,"2"]',
}

export enum ErrorCode {
  BAD_REQUEST = 'BAD_REQUEST',
  NO_RESULT = 'NO_RESULT',
}

export enum RpcIds {
  DEFAULT = 'MkEWBc',
}

export type RcpIdsKeys = keyof typeof RpcIds
