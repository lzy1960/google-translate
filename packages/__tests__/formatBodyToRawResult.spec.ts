import { describe, it, expect } from 'vitest';

import { ErrorCode } from '../../types/index';
import {
  getTranslateData,
  formatBodyToRawResult,
  getResult,
} from '../src/translate';

describe('formatBodyToRawResult', async () => {
  const snapFn = async (text: string) => {
    const res = await getTranslateData(text);
    const data = formatBodyToRawResult(res);
    expect(data).toMatchSnapshot();
  };
  it('should fetch google api and pick up translate used array', async () => {
    const res = await getTranslateData('你好');
    const result = formatBodyToRawResult(res);
    expect(result).toMatchSnapshot();
  });
  it('should translate one word from zh-CN to en', async () => {
    const res = await getTranslateData('你好');
    const data = formatBodyToRawResult(res);
    const result = getResult(data, { from: 'zh-CN', to: 'en' });
    expect(result.text).toBe('Hello');
  });
  it('should get translate data without char', async () => {
    await snapFn('');
  });
  it('should get translate data with space', async () => {
    await snapFn(' ');
  });
  it('should get translate data with punctuation', async () => {
    await snapFn('今天，我看到一个程序员！');
  });
  it('should get translate data with confusing punctuation', async () => {
    await snapFn('这,...是,..什.么???？12123123');
  });
  it('should get translate data with long sentences', async () => {
    const res = await getTranslateData(
      '床前明月光，疑是地上霜。举头望明月，低头思故乡。',
    );
    const data = formatBodyToRawResult(res);
    expect(data).toMatchSnapshot();
  });
  it('should throw error when bad request', async () => {
    const fn = getTranslateData('你好', {
      from: 'auto',
      to: 'en',
      tld: 'error',
    });
    await expect(fn).rejects.toThrowError(ErrorCode.BAD_REQUEST);
  });
});
