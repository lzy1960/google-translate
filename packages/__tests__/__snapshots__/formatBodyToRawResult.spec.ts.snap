// Vitest Snapshot v1

exports[`formatBodyToRawResult > should fetch google api and pick up translate used array 1`] = `
[
  [
    "Nǐ hǎo",
    null,
    "zh-CN",
    [
      [
        [
          0,
          [
            [
              [
                null,
                2,
              ],
            ],
            [
              true,
            ],
          ],
        ],
      ],
      2,
    ],
    [
      [
        "你好",
        null,
        null,
        2,
      ],
    ],
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [
            "Hello",
            null,
            null,
            null,
            [
              [
                "Hello",
                [
                  4,
                  5,
                  11,
                ],
              ],
              [
                "Hi",
                [
                  4,
                ],
              ],
              [
                "Hello there",
                [
                  4,
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      "你好",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
  [
    "你好!",
    null,
    null,
    null,
    null,
    [
      [
        [
          "感叹词",
          [
            [
              "Hello!",
              null,
              [
                "你好!",
                "喂!",
              ],
              1,
              true,
            ],
            [
              "Hi!",
              null,
              [
                "嗨!",
                "你好!",
              ],
              1,
              true,
            ],
            [
              "Hallo!",
              null,
              [
                "你好!",
              ],
              3,
              true,
            ],
          ],
          "en",
          "zh-CN",
        ],
      ],
      3,
    ],
    null,
    null,
    "zh-CN",
    1,
  ],
]
`;

exports[`formatBodyToRawResult > should get translate data with confusing punctuation 1`] = `
[
  [
    "Zhè,... Shì,.. Shén. Me????12123123",
    null,
    "zh-CN",
    [
      [
        [
          0,
          [
            [
              [
                null,
                16,
              ],
            ],
            [
              true,
            ],
          ],
        ],
      ],
      24,
    ],
    [
      [
        "这,...是,..什.么???？",
        null,
        null,
        16,
      ],
      [
        "12123123",
        null,
        16,
        24,
      ],
    ],
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [
            "what is this????",
            null,
            null,
            null,
            [
              [
                "what is this????",
                [
                  2,
                ],
              ],
              [
                "What's this????",
                [
                  4,
                ],
              ],
              [
                "This, ... yes, .. ??? ????",
                [
                  5,
                ],
              ],
              [
                "This, ... yes, .. ??????",
                [
                  11,
                ],
              ],
            ],
          ],
          [
            "12123123",
            null,
            true,
            null,
            [
              [
                "12123123",
                [
                  5,
                ],
              ],
              [
                "1212313",
                [
                  11,
                ],
              ],
            ],
          ],
        ],
        "zh-CN",
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      "这,...是,..什.么???？12123123",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
]
`;

exports[`formatBodyToRawResult > should get translate data with confusing punctuation 2`] = `
[
  [
    "Zhè,... Shì,.. Shén. Me????12123123",
    null,
    "zh-CN",
    [
      [
        [
          0,
          [
            [
              [
                null,
                16,
              ],
            ],
            [
              true,
            ],
          ],
        ],
      ],
      24,
    ],
    [
      [
        "这,...是,..什.么???？",
        null,
        null,
        16,
      ],
      [
        "12123123",
        null,
        16,
        24,
      ],
    ],
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [
            "what is this????",
            null,
            null,
            null,
            [
              [
                "what is this????",
                [
                  2,
                ],
              ],
              [
                "What's this????",
                [
                  4,
                ],
              ],
              [
                "This, ... yes, .. ??? ????",
                [
                  5,
                ],
              ],
              [
                "This, ... yes, .. ??????",
                [
                  11,
                ],
              ],
            ],
          ],
          [
            "12123123",
            null,
            true,
            null,
            [
              [
                "12123123",
                [
                  5,
                ],
              ],
              [
                "1212313",
                [
                  11,
                ],
              ],
            ],
          ],
        ],
        "zh-CN",
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      "这,...是,..什.么???？12123123",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
]
`;

exports[`formatBodyToRawResult > should get translate data with long sentences 1`] = `
[
  [
    "Chuáng qián míngyuèguāng, yí shì dìshàng shuāng. Jǔ tóu wàng míng yuè, dītóu sī gùxiāng.",
    null,
    "zh-CN",
    [
      [
        [
          0,
          [
            [
              [
                null,
                12,
              ],
            ],
            [
              true,
            ],
          ],
        ],
      ],
      24,
    ],
    [
      [
        "床前明月光，疑是地上霜。",
        null,
        null,
        12,
      ],
      [
        "举头望明月，低头思故乡。",
        null,
        12,
        24,
      ],
    ],
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [
            "The bright moonlight in front of the bed was suspected to be frost on the ground.",
            null,
            null,
            null,
            [
              [
                "The bright moonlight in front of the bed was suspected to be frost on the ground.",
                [
                  5,
                ],
              ],
              [
                "The bright moonlight in front of the bed seemed to be frost on the ground.",
                [
                  11,
                ],
              ],
            ],
          ],
          [
            "Raise your head to look at the bright moon, and bow your head to think of your hometown.",
            null,
            true,
            null,
            [
              [
                "Raise your head to look at the bright moon, and bow your head to think of your hometown.",
                [
                  5,
                ],
              ],
              [
                "Raise your head to look at the bright moon, bow your head to think of your hometown.",
                [
                  11,
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      "床前明月光，疑是地上霜。举头望明月，低头思故乡。",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
]
`;

exports[`formatBodyToRawResult > should get translate data with punctuation 1`] = `
[
  [
    "Jīntiān, wǒ kàn dào yīgè chéngxù yuán!",
    null,
    "zh-CN",
    [
      [
        [
          0,
          [
            [
              [
                null,
                12,
              ],
            ],
            [
              true,
            ],
          ],
        ],
      ],
      12,
    ],
    [
      [
        "今天，我看到一个程序员！",
        null,
        null,
        12,
      ],
    ],
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [
            "Today, I see a programmer!",
            null,
            null,
            null,
            [
              [
                "Today, I see a programmer!",
                [
                  5,
                ],
              ],
              [
                "Today, I saw a programmer!",
                [
                  11,
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      "今天，我看到一个程序员！",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
]
`;

exports[`formatBodyToRawResult > should get translate data with punctuation 2`] = `
[
  [
    "Jīntiān, wǒ kàn dào yīgè chéngxù yuán!",
    null,
    "zh-CN",
    [
      [
        [
          0,
          [
            [
              [
                null,
                12,
              ],
            ],
            [
              true,
            ],
          ],
        ],
      ],
      12,
    ],
    [
      [
        "今天，我看到一个程序员！",
        null,
        null,
        12,
      ],
    ],
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [
            "Today, I see a programmer!",
            null,
            null,
            null,
            [
              [
                "Today, I see a programmer!",
                [
                  5,
                ],
              ],
              [
                "Today, I saw a programmer!",
                [
                  11,
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      "今天，我看到一个程序员！",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
]
`;

exports[`formatBodyToRawResult > should get translate data with space 1`] = `
[
  [
    null,
    null,
    "zh-CN",
  ],
  [
    [
      [
        null,
        null,
        null,
        null,
        null,
        [
          [],
        ],
      ],
    ],
    "en",
    1,
    "zh-CN",
    [
      " ",
      "auto",
      "en",
      true,
    ],
  ],
  "zh-CN",
]
`;

exports[`formatBodyToRawResult > should get translate data without char 1`] = `null`;
